### PR TITLE
Implemented os_utime

### PR DESCRIPTION
**Pull request by cklein**
To replace the code in fileio.c.
Fun fact: libuv does not use utimes yet (if available)

### CODE REVIEW

**Comment by josephholsten** [comment]
@aktau did review come to a conclusion with this patch? is this worth updating so it can be merged?

**Comment by aktau** [comment]
> @aktau did review come to a conclusion with this patch? is this worth updating so it can be merged?

What a blast from the past! Thanks for pinging me (I'm actually surprised this successfully reached my inbox, my github mail filters are quite whimsical, for lack of a better word.)

It seems the review got stuck on the prototype of the function. This is the current state AFAICS:

neovim at HEAD:

```c
#ifdef UNIX
static void 
set_file_time (
    char_u *fname,
    time_t atime,               /* access time */
    time_t mtime               /* modification time */
)
/* ... */
#endif
```

This PR:

```c
/// Set access time and modification time of a file
///
/// @return `0` on success, a libuv error code on failure.
int os_utime(const char *path, uv_timespec_t ts_atim, uv_timespec_t ts_mtim)
```

My suggestion:

```c
/// ...
/// @param path ...
/// @param atime the desired access time in nanoseconds since the UNIX epoch
/// @param mtime the desired modification time in nanoseconds since the UNIX epoch
/// @return either 0 for success, or a negative value representing the error (errno code)
int os_utime(const char *path, uint64_t atime, uint64_t mtime) FUNC_ATTR_NONNULL_ALL
```

(With the reasoning I made above.)

There was also the suggestion by @splinterofchaos to implement `os_fileinfo_utime`. Which would shorten the likely call-sites, at the expense of passing much more information than required (an entire fileinfo struct).

Conclusion
-----------

After giving this some thought, taking into account that the current call-sites are not legion (...the opposite), I think it's fine to pass `uv_timespec_t`'s instead of `uint64_t`. Otherwise we need to define another inline function to convert `uv_timespec_t` to `uint64_t`, which will only increase code size. 

When/if somebody comes along who finds that they need to convert a perfectly fine nanosecond `uint64_t` into the gnarly `uv_timespec_t` in order to use `os_utime`, we should change `os_utime` though.

TL;DR: Add `FUNC_ATTR_NONNULL_ALL` to the function prototype to the PR in its current state, rebase and merge. LGTM!

Side note: it would probably be better to use file descriptor based methods whenever possible to avoid race conditions (see [utimes(2)](http://man7.org/linux/man-pages/man2/utime.2.html) vs [futimes(3)](http://man7.org/linux/man-pages/man3/futimes.3.html). But that's far too large a change to make here.

**Comment by justinmk** [comment]
Sorry for missing this. Similar (duplicate :() work was done in https://github.com/neovim/neovim/pull/10357
